
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="NGSPipeDb is a pipeline">
      
      
      
        <meta name="author" content="Xuan Zhang">
      
      
        <link rel="canonical" href="https://xuanblo.github.io/NGSPipeDb/NGSPipeDb/">
      
      <link rel="icon" href="../imgs/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.0.6">
    
    
      
        <title>NGSPipeDb - NGS pipeline and database - NGSPipeDb Document</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.2c0c5eaf.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.7fa14f5b.min.css">
        
          
          
          <meta name="theme-color" content="#7e56c2">
        
      
    
    

<!--<script src="../assets/pymdownx-extras/material-extra-theme-7c147bb7.js" type="text/javascript"></script>-->

    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
      <link rel="stylesheet" href="../stylesheets/extra-aa7dff8003.css">
    
    
      
        
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-117843700-2","auto"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){var e;this.value&&(e=document.location.pathname,ga("send","pageview",e+"?q="+this.value))}),"undefined"!=typeof location$&&location$.subscribe(function(e){ga("send","pageview",e.pathname)})})</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="dracula" data-md-color-primary="deep-purple" data-md-color-accent="deep-purple">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#ngspipedb-ngs-pipeline-and-database" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav
    class="md-header__inner md-grid"
    aria-label="Header"
  >
    <a
      href="https://xuanblo.github.io/NGSPipeDb/"
      title="NGSPipeDb Document"
      class="md-header__button md-logo"
      aria-label="NGSPipeDb Document"
    >
      
  <img src="../imgs/logo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NGSPipeDb Document
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              NGSPipeDb - NGS pipeline and database
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      <div class="md-header-nav__scheme md-header-nav__button md-source__icon md-icon">
          <a
            href="javascript:toggleScheme();"
            title="Light mode"
            class="light-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7l2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10l1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10l-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="Dark mode"
            class="dark-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17.75 4.09l-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91l-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31z"/></svg>
          </a>
          <a
            href="javascript:toggleScheme();"
            title="System preference"
            class="system-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5l1.43 1.43L4.93 20.5 3.5 19.07 19.07 3.5m-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13.51 1.67m-3.3 3.61l-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87.4 1.31M19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58l2.77-1.15-.24 3.35-2.53-2.2m4.33-2.7l1.15-2.77 2.2 2.54-3.35.23m1.15-4.96l-1.14-2.78 3.34.24-2.2 2.54M9.63 18.93l2.77 1.15-2.53 2.19-.24-3.34z"/></svg>
          </a>
           <a
            href="javascript:toggleScheme();"
            title="Unknown scheme"
            class="unknown-mode"
          >
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M15.07 11.25l-.9.92C13.45 12.89 13 13.5 13 15h-2v-.5c0-1.11.45-2.11 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2 2 0 0 0-2-2 2 2 0 0 0-2 2H8a4 4 0 0 1 4-4 4 4 0 0 1 4 4 3.2 3.2 0 0 1-.93 2.25M13 19h-2v-2h2M12 2A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10c0-5.53-4.5-10-10-10z"/></svg>
          </a>
      </div>
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/xuanblo/NGSPipeDb/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-tabs__inner md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  <li class="md-tabs__item">
    <a href=".." class="md-tabs__link">
      Home
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../Reference-based-NGSPipe-RNA-seq/" class="md-tabs__link">
        Pipelines
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../NGSDb/" class="md-tabs__link">
        Databases
      </a>
    </li>
  

      
        
  
  


  <li class="md-tabs__item">
    <a href="../changelog/" class="md-tabs__link">
      Change logs
    </a>
  </li>

      
        
  
  


  <li class="md-tabs__item">
    <a href="../todo/" class="md-tabs__link">
      Todo
    </a>
  </li>

      
    </ul>
  </div>
</nav>
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="NGSPipeDb Document" class="md-nav__button md-logo" aria-label="NGSPipeDb Document" data-md-component="logo">
      
  <img src="../imgs/logo.png" alt="logo">

    </a>
    NGSPipeDb Document
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/xuanblo/NGSPipeDb/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" >
      
      <label class="md-nav__link" for="__nav_2">
        Pipelines
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Pipelines" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Pipelines
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../Reference-based-NGSPipe-RNA-seq/" class="md-nav__link">
        NGSPipe Reference-based RNA-seq
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../Reference-free-NGSPipe-RNA-seq/" class="md-nav__link">
        NGSPipe Denovo RNA-seq
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../Reference-based-NGSPipe-lncRNA-analysis/" class="md-nav__link">
        NGSPipe Reference-based lncRNA
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSPipe-ChIP-seq/" class="md-nav__link">
        NGSPipe ChIP-seq
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSPipe-resequecing/" class="md-nav__link">
        NGSPipe resequecing
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSPipe-medicago-tnt1/" class="md-nav__link">
        NGSPipe medicago tnt insertion
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      <label class="md-nav__link" for="__nav_3">
        Databases
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Databases" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Databases
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSDb/" class="md-nav__link">
        NGSDb with genome
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSDb/" class="md-nav__link">
        NGSDb without genome
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../NGSDb/" class="md-nav__link">
        NGSDb with multi genomes
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        Change logs
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../todo/" class="md-nav__link">
        Todo
      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction-to-ngspipedb" class="md-nav__link">
    Introduction to NGSPipeDb 
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#system-requirements" class="md-nav__link">
    System requirements 
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#anatomy-of-a-ngspipedb-project" class="md-nav__link">
    Anatomy of a NGSPipeDb project 
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#basics-an-example-execution-of-rna-seq-analysis-with-test-data" class="md-nav__link">
    Basics: An example execution of RNA-seq analysis with test data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advance-an-example-execution-of-rna-seq-analysis-with-custome-data" class="md-nav__link">
    Advance: An example execution of RNA-seq analysis with custome data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reproducibility" class="md-nav__link">
    Reproducibility 
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#troubleshooting" class="md-nav__link">
    Troubleshooting 
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contributing" class="md-nav__link">
    Contributing
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
  
                
                  <a href="https://github.com/xuanblo/NGSPipeDb/edit/master/mkdocs/markdowns/NGSPipeDb.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="ngspipedb-ngs-pipeline-and-database">NGSPipeDb - NGS pipeline and database</h1>
<p><strong>Author:</strong> Dr. Xuan Zhang<em> <sup><a href="https://github.com/xuanblo"><img alt="github" src="https://img.icons8.com/ios/15/000000/github.png" /></a></sup> <sup><a href="https://scholar.google.com/citations?user=omUk0vUAAAAJ"><img alt="google scholar" src="https://img.icons8.com/material/17/000000/google-scholar--v2.png" /></a></sup><br />
<strong>Last update:</strong> 2021-01-20</em><br />
<strong>Citation:</strong> NGSPipeDb: An automated pipeline for parallel processing of huge NGS data and database generation.</p>
<p><strong>Table of Contents:</strong></p>
<ul>
<li><a href="#Intro">Introduction to NGSPipeDb</a></li>
<li><a href="#Require">System requirements</a></li>
<li><a href="#Anatomy">Anatomy of a NGSPipeDb project</a></li>
<li>
<p><a href="#QuickStarted">Basics: An example execution of RNA-seq analysis with test data</a></p>
</li>
<li>
<p><a href="#QuickStarted">Advance: An example execution of RNA-seq analysis with custome data</a></p>
</li>
<li><a href="#Reproducibility">Reproducibility</a></li>
<li><a href="#Troubleshooting">Troubleshooting</a></li>
</ul>
<h2 id="introduction-to-ngspipedb">Introduction to NGSPipeDb <a name="Intro"></a></h2>
<p><strong>NGSPipeDb</strong> is an automated pipeline for parallel processing of huge next generation sequencing (NGS) data and database generation using <a href="https://snakemake.readthedocs.io/en/stable/index.html">snakemake workflow</a> which allows for ease of use, optimal speed, and a highly modular code that can be further added onto and customized by experienced users. It can be further divided into <code>NGSPipe</code> and <code>NGSDb</code> for individual usage. </p>
<p><strong>NGSPipe</strong> consists of a <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html">Snakefile</a> (<code>ngspipe/rnaseq.snakefile.py</code>, it includes some basic rules <code>ngspipe/rule/*.snakefile.py</code>), <a href="https://conda.io/docs/">conda</a> environment files (<code>ngspipe/envs/*.yaml</code>), a configuration file (<code>ngspipe/config/rnaseq.config.yaml</code>), a set of <a href="#">python</a>, <a href="#">R</a>, <a href="#">Shell</a> and <a href="#">Perl</a> scripts (<code>ngspipe/scripts/*.py</code>), and a set of <a href="#">reStructuretext</a> reports (<code>reports/*.rst</code>). It combines the use of several dozen omic-seq tools, suites, and packages to create a complete pipeline that takes <a href="">RNA-seq analysis</a>, <a href="">resequcing analysis</a> etc. from raw sequencing data all the way through alignment, quality control, unsupervised analyses, differential expression, and downstream pathway analysis. It is implemented such that alternative or similar analysis can be added or removed. The results are compiled in a simple and highly visual <a href="ngspipe/metadata/report.html">report</a> containing the key figures to explain the analysis, and then compiles all of the relevant files, tables, and pictures into an easy to navigate folder. Table file such as csv, tsv, xlsx etc. </p>
<p>It is based on snakemake and includes the following tools:
* shovill (based on Spades)
* QUAST v.5 (including BUSCO)
* mash
* fastp</p>
<p>It will read untrimmed raw data from your illumina sequencing experiments as paired .fastq.gz-files. These are then trimmed, assembled and polished. Besides generating ready-for-use contigs, AQUAMIS will select the closest reference genome from NCBI RefSeq and produce an intuitive, detailed report on your data and assemblies to evaluate its reliability for further analyses. It relies on reference-based and reference-free measures such as coverage depth, gene content, genome completeness and contamination, assembly length and many more. Based on the experience from thousands of sequencing experiments, threshold sets for different species have been defined to detect potentially poor results.</p>
<p>In addition, <strong>NGSDb</strong> has been outfitted with several recently published tools that allow for visualize and data share.can be convert to <a href="#">Sqlite3</a> format. The <a href="#">Django</a> project and apps can be orgined by user defined. It is easy to share your data with a web inteface. a set of <code>apps</code> (such as <code>home</code>, <code>igv</code>, <code>geneExpAtlas</code>, <code>efp brwose</code>).</p>
<p>By default, the <strong>NGSPipeDb</strong> performs all the steps shown in the <a href="img/report_2019_03_14_salmonAlignment_visualization.png">diagram</a> below. However, advanced user, you can easily modify the <code>Snakefile</code> and the <code>config.yaml</code> and/or add "custom rules" to enable additional functions.</p>
<p><img alt="img" src="../imgs/workflow.png" /></p>
<p>Currently, transcript quantification with <code>Salmon</code> at the read-level or gene quantification by <a href="http://subread.sourceforge.net"><code>featureCounts</code></a> can be activated.
The first version handles <a href="#">RNA-Seq</a> workflow.</p>
<p>Workflows available:
- RNA-seq
- ChIP-seq
- Resequencing</p>
<p><strong>TODO</strong>:</p>
<ul>
<li>NGSPipe<ol>
<li>miRNA</li>
<li>scRNA-seq</li>
<li>ATAC-seq</li>
</ol>
</li>
<li>NGSdb<ol>
<li>efp browser</li>
</ol>
</li>
</ul>
<h2 id="system-requirements">System requirements <a name="Require"></a></h2>
<p>Building NGSPipeDb and running the examples require Linux, MacOS or Windows Subsystem for Linux (<a href="https://wiki.ubuntu.com/WSL">WSL</a>) on Win10. Other Unix environments will probably work but have not been tested.</p>
<p>The test data can be run on personal computer, for example 8G memeory.</p>
<p>Some of the tools that NGSPipeDb uses, e.g. STAR and cufflinks are very memory intensive programs. Therefore we recommend the following system requirements for NGSPipeDb:</p>
<p>We recommend that you run NGSPipeDb on a server that has at least 30GB of ram. This will allow for a single-threaded NGSPipeDb run (on mouse samples).</p>
<p>We recommend that you have at least 128GB of ram and at least a 4-core CPU if you want to run NGSPipeDb in multi-threaded mode (which will speedup the workflow significantly). Our own servers have 256GB of ram and 32 cores.</p>
<h2 id="anatomy-of-a-ngspipedb-project">Anatomy of a NGSPipeDb project <a name="Anatomy"></a></h2>
<p>It is recommended to download NGSPipeDb source and change its name to your project name (For example: <code>mv NGSPipeDb mouse_transcriptome_analysis</code>), it may looks like the following structure (command: <code>tree -d -L 2 mouse_transcriptome_analysis</code>):</p>
<div class="highlight"><pre><span></span><code>mouse_transcriptome_analysis
├── README.md
├── ngsdb
│   ├── blastplus
│   ├── db.sqlite3
│   ├── geneAnno
│   ├── geneExpAtlas
│   ├── home
│   ├── igv
│   ├── manage.py
│   └── ngsdb
├── ngspipe
│   ├── config
│   ├── db_generate.Snakefile.py
│   ├── envs
│   ├── imgs
│   ├── notebooks
│   ├── reports
│   ├── rnaseq_analysis.Snakefile.py
│   ├── rules
│   └── scripts
├── results
│   ├── report
│   ├── resultdata
│   └── sqlite3
└── testdata
</code></pre></div>
<p>The workflow code goes into a subfolder <code>ngspipe</code>, while the configuration is stored in a subfolder <code>config</code>. Inside of the workflow subfolder, the central Snakefile marks the entrypoint of the workflow. In addition to the central Snakefile, rules are stored in a modular way, using the optional subfolder <code>ngspipe/rules</code>. Further, scripts are stored in a subfolder <code>workflow/scripts</code> and notebooks in a subfolder <code>workflow/notebooks</code>. Conda environments are stored in a subfolder <code>workflow/envs</code>. Finally, report caption files are stored in <code>workflow/report</code>. </p>
<p>The database code goes into a subfolder <code>ngsdb</code>, while the <code>manage.py</code> is ngsdb's command-line utility for administrative tasks. A golabl setting file is stored under <code>ngsdb/ngsdb</code>, such as <code>ngsdb/ngsdb/setting.py</code> and <code>ngsdb/ngsdb/urls.py</code>. Many ngsdb function module take a app name. For example, if your INSTALLED_APPS in <code>ngsdb/ngsdb/setting.py</code> contains the string 'igv', the database will contain a page of IGV genome browser.</p>
<p>All output files generated in the workflow should be stored under <code>results/result</code>, unless they are rather retrieved report, in which case they should be stored under <code>results/report</code>. The latter subfolder <code>results/sqlite3</code> contains Sqlite3 kind file that shall be used by ngsdb.</p>
<h2 id="basics-an-example-execution-of-rna-seq-analysis-with-test-data">Basics: An example execution of RNA-seq analysis with test data</h2>
<h2 id="advance-an-example-execution-of-rna-seq-analysis-with-custome-data">Advance: An example execution of RNA-seq analysis with custome data</h2>
<h2 id="reproducibility">Reproducibility <a name="Reproducibility"></a></h2>
<p>conda环境克隆conda create -n ngspipedb_py38_conda_env &ndash;clone ./ngspipedb_py38_conda_env/</p>
<ol>
<li>use conda env export</li>
</ol>
<div class="highlight"><pre><span></span><code><span class="nb">cd</span> NGSPipeDB_source_code
<span class="c1"># export to yaml</span>
conda env <span class="nb">export</span> --no-builds -p ./ngspipedb_py38_conda_env &gt;ngspipedb_py38_conda_env.yaml
</code></pre></div>
<ol>
<li>use conda pack</li>
</ol>
<p>用&ndash;use-conda这个参数的话，因为所有软件的环境都是单独的，所有conda安装的时候不会出错，那么如果已经下载安装好了环境，用这种方式如何使用？默认的环境是.snakemake文件夹下，如何指定？
用上面的方式好安装，不会出错，但是会导致文件很大，多大？
是否能把一环境分成两部分？一部分软件集合起来变成一个大环境，另一部分软件就用&ndash;use-conda环境单独指定，但是这两种方式能结合到一起用吗？</p>
<div class="highlight"><pre><span></span><code><span class="c1"># pack</span>
<span class="nb">cd</span> NGSPipeDB_source_code
mamba install -c conda-forge conda-pack
conda pack -p ./ngspipedb_py38_conda_env -o ngspipedb_py38_conda_env_osx64.tar.gz
<span class="c1"># unpack on another machine</span>
mkdir -p ngspipedb_py38_conda_env
tar -xzf ngspipedb_py38_conda_env_osx64.tar.gz -C ngspipedb_py38_conda_env
<span class="nb">source</span> activate ./ngspipedb_py38_conda_env
conda-unpack
</code></pre></div>
<p>conda config &ndash;add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free/
conda config &ndash;add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/main/
conda config &ndash;add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge/
conda config &ndash;add channels https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/bioconda/</p>
<ol>
<li>
<p>activate base and set miniconda path
conda init</p>
</li>
<li>
<p>Conda Prompt Customization
conda config &ndash;set env_prompt '({name}) '</p>
</li>
</ol>
<p>source ~/.bashrc</p>
<ol>
<li>
<p>update conda, (optional)
conda update conda</p>
</li>
<li>
<p>create conda visual environment, python version, snakemake version, env directory,django version
conda create -p ngspipedb_py38_conda_env python=3.8</p>
</li>
<li>
<p>activate conda env
conda activate ./ngspipedb_py38_conda_env</p>
</li>
<li>
<p>install mamba to make install software faster.
conda install mamba -c conda-forge</p>
</li>
<li>
<p>update some bioinformatics tools we will use bellow.
mamba env update &ndash;prefix ./ngspipedb_py38_conda_env/ &ndash;file requirement.yaml  &ndash;prune</p>
</li>
<li>
<p>you can exit virtual environment by 
conda deactivate
https://wooey.readthedocs.io/en/latest/install.html</p>
</li>
</ol>
<p>&ndash;conda-frontend mamba 选择更快一点的mamba</p>
<p>&ndash;conda-create-envs-only 只创建环境，然后退出，不运行程序，这个可以用来专门测试环境</p>
<p>mac上的conda环境好像没有linux上面那么好用，特别是anaconda创建的环境
&ndash;conda-prefix 指定conda环境安装地址
清理conda安装包和缓存
snakemake -s ngspipe/db_generate.Snakefile.py &ndash;use-conda &ndash;conda-prefix condaEnvSplit -p -j1</p>
<p>Simplest is just abandon the &ndash;use-conda flag, as suggested in the answer. Alternatively, you could make a container that has the env pre-created and configured, then use &ndash;use-singularity. Or, if the post-installation can be automated, one could build a custom Conda package that runs some post-linking scripts. Sorry I seem to have missed your comment!</p>
<p>snakemake 如何运行单个程序？这个也很有用
基因的命令，像dkango这样的命令在很多rules中都有，所有比如有个顶层的环境中安装了django</p>
<h2 id="troubleshooting">Troubleshooting <a name="Troubleshooting"></a></h2>
<p>Ngsdb.yaml+wooey
Python=3.8 samtools clustergrammer
Pip install wooey
pip install pandas==0.25.3</p>
<h2 id="contributing">Contributing</h2>
<p>Please <a href="https://github.com/xuanblo/NGSPipeDb/issues">submit an issue</a> to report bugs or ask questions.</p>
<p>Please contribute bug fixes or new features with a <a href="https://github.com/xuanblo/NGSPipeDb/pulls">pull request</a> to this
repository.</p>
<p>If this does not help, please feel free to consult:
* Xuan Zhang (<a href="mailto:zhangxuan@xtbg.ac.cn">zhangxuan@xtbg.ac.cn</a>) or</p>
                
              
  <!--
  <footer class="sponsorship">
    <hr>
    <a href="https://github.com/sponsors/facelessuser" title="Become a sponsor">
      <span class="twemoji heart-throb-hover">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><path fill-rule="evenodd" d="M7.655 14.916L8 14.25l.345.666a.752.752 0 0 1-.69 0zm0 0L8 14.25l.345.666.002-.001.006-.003.018-.01a7.643 7.643 0 0 0 .31-.17 22.08 22.08 0 0 0 3.433-2.414C13.956 10.731 16 8.35 16 5.5 16 2.836 13.914 1 11.75 1 10.203 1 8.847 1.802 8 3.02 7.153 1.802 5.797 1 4.25 1 2.086 1 0 2.836 0 5.5c0 2.85 2.045 5.231 3.885 6.818a22.075 22.075 0 0 0 3.744 2.584l.018.01.006.003h.002z"/></svg>
      </span>
    </a>
    <hr>
  </footer>
  -->

              
                


              
            </article>
          </div>
        </div>
      </main>
      
        

<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            Copyright © 2020 - 2021 Xuan Zhang
          </div>
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        with emoji by
        <a href="https://github.com/twitter/twemoji" target="_blank" rel="noopener">
          Twemoji
        </a>
      </div>
      
  <div class="md-footer-social">
    
      
      
        
        
      
      <a href="https://github.com/xuanblo" target="_blank" rel="noopener" title="github.com" class="md-footer-social__link">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
      </a>
    
  </div>

    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "search.highlight"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../assets/javascripts/workers/search.fb4a9340.min.js", "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.a1c7c35e.min.js"></script>
      
        <script src="../javascripts/material-extra-theme-7c147bb7.js"></script>
      
        <script src="../javascripts/extra-uml-51b11a9a.js"></script>
      
    
  </body>
</html>